### ISSUE TITLE
Please add support for single middle mouse click for 'SecondaryActivate'

### ISSUE THREAD

**Issue by guzzard** [ubuntu/gnome-shell-extension-appindicator#89]
In Unity it's possible to single middle mouse click an appindicator icon to run the 'SecondaryActivate' function (if used by the application). Please add this support for this extension (without opening the appindicator menu).

Further info:
https://bugs.launchpad.net/unity/+bug/812933
https://developer.ubuntu.com/api/devel/ubuntu-13.10/c/AppIndicator3-0.1.html

**Comment by jhasse** [comment]
Do you know an application which uses this?

**Comment by guzzard** [comment]
hi @jhasse,

In [radiotray-ng](https://github.com/ebruck/radiotray-ng/issues/25#issuecomment-329899121) v0.1.8 it will be used to turn on/off currently selected radio station. This is also how the original [radiotray](https://bitbucket.org/carlmig/radio-tray/src/c0e5087963c5cf6332fbb62d3cc9b2e4c270562c/src/SysTrayGui.py?at=default&fileviewer=file-view-default#SysTrayGui.py-125) works.

I'm sure there are other applications that use it for quick activation/deactivation of things but radiotray is what I use it for.

**Comment by jhasse** [comment]
As you said in https://github.com/ebruck/radiotray-ng/issues/25#issuecomment-329899784, the problem will be how to close the menu after the click.

**Comment by guzzard** [comment]
Yeah, I did a quick test to see if I could implement it but I'm not that familiar with gjs/JavaScript unfortunately. Capturing the middle mouse click is no problem (you already do it for detecting double clicking), and triggering the 'SecondaryActivate' dbus call is no problem either. The problem is that the appindicator menu opens up regardless of what mouse button is clicked, so when single middle mouse clicking the appindicator menu remains open which is a kind of ugly solution.

I think it might be possible to get it working by creating a custom [Button class](https://github.com/GNOME/gnome-shell/blob/master/js/ui/panelMenu.js#L92) and implement a custom _onEvent function that distinguishes between the different mouse buttons. The _onEvent function in the default Button class does not.

```javascript
    _onEvent: function(actor, event) {
        if (this.menu &&
            (event.type() == Clutter.EventType.TOUCH_BEGIN ||
             event.type() == Clutter.EventType.BUTTON_PRESS))
            this.menu.toggle();

        return Clutter.EVENT_PROPAGATE;
    },
```

[Here](https://blog.fpmurphy.com/2012/01/dual-action-menu-for-gnome-shell-3-2-button.html) is an example of how to get different menus for left and right mouse button. It's for GNOME Shell 3.4, and does not work in 3.24/3.26. However, it should be possible to take those ideas and implement to get a single middle mouse button click working I think.

**Comment by lestcape** [comment]
Well, i just will throw some ideas here. I hope this help.

In theory, all menus of gnome shell are associate with the `menumanager` of the `panel` https://github.com/GNOME/gnome-shell/blob/master/js/ui/panel.js#L97

And you can access the menumanager using `Main.panel._menuManager` The menumanager can give you the current active menu: https://github.com/GNOME/gnome-shell/blob/master/js/ui/popupMenu.js#L1268 witch in theory if have one active is the currently opened menu and you can close it.

A little more general method that won't supposed that we are using the panel `menumanger` and won't supposed that the current active menu is the active menu of the menu manager, it will iterate for all children in Main.uiGroup https://github.com/GNOME/gnome-shell/blob/master/js/ui/main.js#L76 where is supposed that the actor of the menu was set as a children and also have a `_delegate` property that point to the class of the actor, with give you a way to filter only actors that belong to a PopupMenu class: https://github.com/GNOME/gnome-shell/blob/master/js/ui/popupMenu.js#L77

So we can detect and close the active menu also in this way asking if the menu is open https://github.com/GNOME/gnome-shell/blob/master/js/ui/popupMenu.js#L431

Also in theory gnome shell only can have one opened menu at the same time and this fact can used to optimized the search or not if is more important be more general.

**Comment by guzzard** [comment]
@lestcape thanks for the input! I got it working beautifully thanks to your info! :+1: 

@jhasse pull request coming soon.

**Comment by jhasse** [comment]
fixed by #90.
